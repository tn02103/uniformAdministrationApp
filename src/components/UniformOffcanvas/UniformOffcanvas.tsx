import { deleteUniformItem } from "@/dal/uniform/item/_index";
import { AuthRole } from "@/lib/AuthRoles";
import { useI18n } from "@/lib/locales/client";
import { Uniform, UniformType } from "@/types/globalUniformTypes";
import { UniformFormType } from "@/zod/uniform";
import { useState } from "react";
import { Offcanvas, Row } from "react-bootstrap";
import { toast } from "react-toastify";
import { LabelIconButton } from "../Buttons/LabelIconButton";
import { useGlobalData } from "../globalDataProvider";
import { useModal } from "../modals/modalProvider";
import { UniformDeficiencyRow } from "./UniformDeficiencyRow";
import { UniformDetailRow } from "./UniformDetailRow";
import { UniformHistoryRow } from "./UniformHistoryRow";

export type UniformOffcanvasProps = {
    uniform: Uniform;
    uniformType: UniformType;
    onClose: () => void;
    onSave: (data?: UniformFormType) => void;
}
export const UniformOffcanvas = ({ uniform, uniformType, onClose, onSave }: UniformOffcanvasProps) => {
    const t = useI18n();
    const [editable, setEditable] = useState(false);
    const modal = useModal();
    const { userRole } = useGlobalData();

    const handleDelete = () => {
        const deleteAction = () => deleteUniformItem(uniform.id)
            .then(() => {
                toast.success(t('uniformOffcanvas.deleteAction.success'));
                onSave();
                onClose();
            }).catch(() =>
                toast.error(t('uniformOffcanvas.deleteAction.failed'))
            );

        modal?.simpleWarningModal({
            header: t('uniformOffcanvas.deleteAction.header', { type: uniform.type.name, number: uniform.number }),
            message: (
                <p>
                    {t('uniformOffcanvas.deleteAction.message.one', { type: uniform.type.name, number: uniform.number })}<br />
                    {t('uniformOffcanvas.deleteAction.message.two')}
                </p>
            ),
            primaryOption: t('common.actions.delete'),
            type: 'danger',
            primaryFunction: deleteAction,
        });
    }

    return (
        <Offcanvas
            show={true}
            onHide={onClose}
            placement="end"
            backdrop={true}
            scroll={false}
            style={{ width: "576px" }}
            aria-labelledby="offcanvas-uniform-header"
        >
            <Offcanvas.Header closeButton>
                <Offcanvas.Title>
                    <h3 id="offcanvas-uniform-header">
                        {uniform.type.name}: <b>{uniform.number}</b>
                    </h3>
                </Offcanvas.Title>
            </Offcanvas.Header>
            <Offcanvas.Body>
                <h4>{t('common.details')}</h4>
                <hr className="mb-0" />
                {userRole > AuthRole.user && (
                    <Row className="justify-content-evenly">
                        <LabelIconButton
                            variantKey="edit"
                            disabled={editable}
                            onClick={() => setEditable(true)}
                        />
                        <LabelIconButton
                            variantKey="delete"
                            disabled={editable}
                            onClick={handleDelete}
                        />
                    </Row>
                )}
                <UniformDetailRow
                    uniform={uniform}
                    uniformType={uniformType}
                    editable={editable}
                    setEditable={() => setEditable(!editable)}
                    onSave={onSave}
                />
                {userRole > AuthRole.user && (
                    <>
                        <h4>{t('uniformOffcanvas.deficiency.header')}</h4>
                        <hr className="mb-0" />
                        <UniformDeficiencyRow uniformId={uniform.id} />
                        <h4 className="mt-4">{t('uniformOffcanvas.history.header')}</h4>
                        <hr className="mb-0" />
                        <UniformHistoryRow uniformId={uniform.id} />
                    </>
                )}
            </Offcanvas.Body>
        </Offcanvas>
    );
}
