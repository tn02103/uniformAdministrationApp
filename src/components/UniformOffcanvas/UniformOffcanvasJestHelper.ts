import { Deficiency } from "@/types/deficiencyTypes";
import { generationLists, sizeLists, typeList } from "../../../tests/_jestConfig/staticMockData";
import { AuthRole } from "@/lib/AuthRoles";


// ------------- MOCKS FOR DEFICIENCY -------------
jest.mock('@/dal/inspection/deficiency', () => ({
    createUniformDeficiency: jest.fn(async () => "created successfully"),
    resolveDeficiency: jest.fn(async () => "resolved successfully"),
    updateUniformDeficiency: jest.fn(async () => "updated successfully"),
}));

jest.mock('@/dataFetcher/deficiency', () => ({
    useDeficienciesByUniformId: jest.fn((_, includeResolved) => {
        if (includeResolved) {
            return { deficiencies: mockDeficiencyList };
        } else {
            return { deficiencies: mockDeficiencyList.slice(0, 2) };
        }
    }),
    useDeficiencyTypes: jest.fn(() => ({ deficiencyTypeList: mockDeficiencyTypeList })),
}));

// ------------- MOCKS FOR UNIFORM -------------

jest.mock('@/dataFetcher/uniformAdmin', () => ({
    useUniformGenerationListByType: jest.fn(() => ({
        generationList: generationLists[0]
    })),
    useUniformTypeList: jest.fn(() => ({
        typeList
    })),
}));
jest.mock("@/dal/uniform/item/_index", () => ({
    updateUniformItem: jest.fn(() => Promise.resolve('Saved item')),
    deleteUniformItem: jest.fn(() => Promise.resolve('Deleted item')),
    getUniformItemHistory: jest.fn(() => Promise.resolve([])), // Mock for history
    getUniformItemCountByType: jest.fn(() => Promise.resolve(0)), // Mock for count
    issueUniformItem: jest.fn(),
    createUniformItems: jest.fn(),
    getUniformItemDeficiencies: jest.fn(),
}));
jest.mock('../globalDataProvider', () => ({
    useGlobalData: jest.fn(() => ({
        sizelists: sizeLists,
        userRole: global.__ROLE__ ?? AuthRole.admin,
    })),
}));
jest.mock('@/dataFetcher/uniform', () => ({
    useUniformItemHistory: jest.fn(() => ({
        history: mockUniformHistory,
    })),
}));

// ------------- OTHER MOCKS -------------
jest.mock("swr", () => ({
    mutate: jest.fn(async () => { }),
}));

jest.mock('react-toastify', () => ({
    toast: {
        error: jest.fn(),
        success: jest.fn(),
    },
}));

export const mockDeficiencyTypeList = [
    {
        id: 'de3860d4-c88e-4a7c-be4c-7e832eda31d4',
        name: 'Test Type',
        dependent: 'uniform',
    },
    {
        id: '61d28738-4757-4e98-bfbf-a82f8404a74a',
        name: 'Another One',
        dependent: 'uniform',
    },
];
export const mockDeficiencyList: Deficiency[] = [
    {
        id: '80c36a66-81e9-4c9c-8d56-a095a6e9e28b',
        comment: 'Test comment',
        description: 'Test description',
        dateCreated: new Date('2023-10-01T12:00:00Z'),
        dateUpdated: new Date('2023-10-01T12:00:00Z'),
        userCreated: 'user',
        userUpdated: 'user',
        typeId: mockDeficiencyTypeList[0].id,
        typeName: mockDeficiencyTypeList[0].name,
    },
    {
        id: 'a1b2c3d4-e5f6-7g8h-9i0j-k1l2m3n4o5p6',
        comment: 'Test comment 2',
        description: 'Test description 2',
        dateCreated: new Date('2023-10-02T12:00:00Z'),
        dateUpdated: new Date('2023-10-12T12:00:00Z'),
        userCreated: 'user',
        userUpdated: 'user1',
        typeId: mockDeficiencyTypeList[1].id,
        typeName: mockDeficiencyTypeList[1].name,
    },
    {
        id: 'b7c8d9e0-f1g2-h3i4-j5k6-l7m8n9o0p1q2',
        comment: 'Test comment 3',
        description: 'Test description 3',
        dateCreated: new Date('2023-10-03T12:00:00Z'),
        dateUpdated: new Date('2023-10-12T12:00:00Z'),
        dateResolved: new Date('2023-10-17T12:00:00Z'),
        userCreated: 'user',
        userUpdated: 'user1',
        userResolved: 'user2',
        typeId: mockDeficiencyTypeList[0].id,
        typeName: mockDeficiencyTypeList[0].name,
    },
];
export const mockUniform = {
    id: "c227ac23-93d4-42b5-be2e-956ea35c2db9",
    number: 2501,
    generation: generationLists[0][1],
    size: sizeLists[0].uniformSizes[0],
    comment: "Test comment",
    active: true,
    type: {
        id: typeList[0].id,
        name: typeList[0].name,
    },
};
export const mockUniformHistory = [
    {
        dateIssued: new Date('2023-10-01T12:00:00Z'),
        dateReturned: null,
        cadet: {
            firstname: 'John',
            lastname: 'Doe',
            recdelete: false,
        },
    },
    {
        dateIssued: new Date('2023-10-02T12:00:00Z'),
        dateReturned: new Date('2023-10-03T12:00:00Z'),
        cadet: {
            firstname: 'Jane',
            lastname: 'Smith',
            recdelete: true,
        },
    },
];