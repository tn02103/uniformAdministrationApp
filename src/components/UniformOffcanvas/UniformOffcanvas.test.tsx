import "./UniformOffcanvasJestHelper";

import { AuthRole } from "@/lib/AuthRoles";
import { getAllByRole, getByLabelText, getByText, queryAllByRole, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { typeList } from "../../../tests/_jestConfig/staticMockData";
import { UniformOffcanvas } from "./UniformOffcanvas";
import { mockUniform } from "./UniformOffcanvasJestHelper";

describe('UniformOffcanvas', () => {
    const { useModal } = jest.requireMock('../modals/modalProvider');
    const { simpleWarningModal } = useModal();
    const { deleteUniformItem, updateUniformItem } = jest.requireMock('@/dal/uniform/item/_index');
    const { toast } = jest.requireMock('react-toastify');

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should render correctly', () => {
        render(
            <div>
                <UniformOffcanvas
                    uniform={mockUniform}
                    uniformType={typeList[0]}
                    onClose={jest.fn()}
                    onSave={jest.fn()}
                />
            </div>
        );

        expect(screen.getByRole('dialog')).toMatchSnapshot();
    });

    it('should call onClose when close button is clicked', () => {
        const onCloseMock = jest.fn();
        const { getByLabelText } = render(
            <UniformOffcanvas
                uniform={mockUniform}
                uniformType={typeList[0]}
                onClose={onCloseMock}
                onSave={jest.fn()}
            />
        );

        const closeButton = getByLabelText('Close');
        closeButton.click();

        expect(onCloseMock).toHaveBeenCalledTimes(1);
    });

    it('should set editable if edit button is clicked', async () => {
        const user = userEvent.setup();
        const { getByRole } = render(
            <UniformOffcanvas
                uniform={mockUniform}
                uniformType={typeList[0]}
                onClose={jest.fn()}
                onSave={jest.fn()}
            />
        );

        const editButton = getByRole('button', { name: /edit/i });
        await user.click(editButton);

        expect(editButton).toBeDisabled();
        expect(getByRole('button', { name: /delete/i })).toBeDisabled();
        expect(getByRole('button', { name: /save/i })).toBeEnabled();
        expect(getByRole('button', { name: /cancel/i })).toBeEnabled();

        expect(getByRole('switch', { name: 'common.status' })).toBeEnabled();
        expect(getByRole('combobox', { name: 'common.uniform.size' })).toBeEnabled();
        expect(getByRole('combobox', { name: 'common.uniform.generation.label' })).toBeEnabled();
        expect(getByRole('textbox', { name: /comment/i })).toBeEnabled();
    });

    it('should reset editable if cancel button is clicked', async () => {
        const user = userEvent.setup();
        const { getByRole, queryByRole } = render(
            <UniformOffcanvas
                uniform={mockUniform}
                uniformType={typeList[0]}
                onClose={jest.fn()}
                onSave={jest.fn()}
            />
        );

        const editButton = getByRole('button', { name: /edit/i });
        await user.click(editButton);

        const cancelButton = getByRole('button', { name: /cancel/i });
        await user.click(cancelButton);

        expect(getByRole('button', { name: /edit/i })).toBeEnabled();
        expect(getByRole('button', { name: /delete/i })).toBeEnabled();

        expect(queryByRole('button', { name: /save/i })).toBeNull();
        expect(queryByRole('button', { name: /cancel/i })).toBeNull();
        expect(getByRole('textbox', { name: /comment/i })).toHaveAttribute('disabled');
    });

    describe('history', () => {
        it('should render history', () => {
            render(
                <UniformOffcanvas
                    uniform={mockUniform}
                    uniformType={typeList[0]}
                    onClose={jest.fn()}
                    onSave={jest.fn()}
                />
            );

            const list = screen.getByRole('list', { name: 'uniformOffcanvas.history.header' });
            expect(getAllByRole(list, 'listitem')).toHaveLength(2);

            const firstitem = getAllByRole(list, 'listitem')[0];
            expect(getByLabelText(firstitem, 'dateIssued')).toHaveTextContent('01.10.2023');
            expect(getByLabelText(firstitem, 'dateReturned')).toHaveTextContent('');
            expect(getByLabelText(firstitem, 'person')).toHaveTextContent('John Doe');
            expect(getByLabelText(firstitem, 'person')).not.toHaveClass('text-decoration-line-through');

            const secondItem = getAllByRole(list, 'listitem')[1];
            expect(getByLabelText(secondItem, 'dateIssued')).toHaveTextContent('02.10.2023');
            expect(getByLabelText(secondItem, 'dateReturned')).toHaveTextContent('03.10.2023');
            expect(getByLabelText(secondItem, 'person')).toHaveTextContent('Jane Smith');
            expect(getByLabelText(secondItem, 'person')).toHaveClass('text-decoration-line-through');
        });

        it('should render empty history', () => {
            const { useUniformItemHistory } = jest.requireMock('@/dataFetcher/uniform');
            useUniformItemHistory.mockReturnValue({
                history: [],
            });

            render(
                <UniformOffcanvas
                    uniform={mockUniform}
                    uniformType={typeList[0]}
                    onClose={jest.fn()}
                    onSave={jest.fn()}
                />
            );

            const list = screen.getByRole('list', { name: 'uniformOffcanvas.history.header' });
            expect(queryAllByRole(list, 'listitem')).toHaveLength(0);
            expect(getByText(list, 'uniformOffcanvas.history.noEntries')).toBeInTheDocument();
        });
    });

    describe('dal-methods', () => {
        describe('updateUniformItem', () => {
            it('should update item', async () => {
                const user = userEvent.setup();
                const onSaveMock = jest.fn();
                const { getByRole } = render(
                    <UniformOffcanvas
                        uniform={mockUniform}
                        uniformType={typeList[0]}
                        onClose={jest.fn()}
                        onSave={onSaveMock}
                    />
                );

                const editButton = getByRole('button', { name: /edit/i });
                await user.click(editButton);

                const saveButton = getByRole('button', { name: /save/i });
                await user.click(saveButton);

                expect(updateUniformItem).toHaveBeenCalledTimes(1);
                expect(updateUniformItem).toHaveBeenCalledWith({
                    id: mockUniform.id,
                    number: mockUniform.number,
                    generation: mockUniform.generation.id,
                    size: mockUniform.size.id,
                    comment: mockUniform.comment,
                    active: mockUniform.active,
                });
                expect(onSaveMock).toHaveBeenCalledTimes(1);

                expect(getByRole('button', { name: /edit/i })).toBeEnabled();
                expect(getByRole('button', { name: /delete/i })).toBeEnabled();
                expect(getByRole('textbox', { name: /comment/i })).toHaveAttribute('disabled');
            });
            it('should catch exception', async () => {
                const user = userEvent.setup();
                const onSaveMock = jest.fn();
                updateUniformItem.mockRejectedValueOnce(new Error('Failed to save item'));

                const { getByRole } = render(
                    <UniformOffcanvas
                        uniform={mockUniform}
                        uniformType={typeList[0]}
                        onClose={jest.fn()}
                        onSave={onSaveMock}
                    />
                );

                const editButton = getByRole('button', { name: /edit/i });
                await user.click(editButton);

                const saveButton = getByRole('button', { name: /save/i });
                await user.click(saveButton);

                expect(onSaveMock).toHaveBeenCalledTimes(0);
                expect(toast.error).toHaveBeenCalledTimes(1);
                expect(toast.error).toHaveBeenCalledWith('common.error.actions.save');

                expect(getByRole('button', { name: /edit/i })).toBeDisabled();
                expect(getByRole('button', { name: /delete/i })).toBeDisabled();
                expect(getByRole('button', { name: /save/i })).toBeInTheDocument();
                expect(getByRole('textbox', { name: /comment/i })).not.toHaveAttribute('disabled');
            });
        });

        describe('deleteUniformItem', () => {
            it('should delete item', async () => {
                const onCloseMock = jest.fn();
                const onSaveMock = jest.fn();

                const user = userEvent.setup();
                const { getByRole } = render(
                    <UniformOffcanvas
                        uniform={mockUniform}
                        uniformType={typeList[0]}
                        onClose={onCloseMock}
                        onSave={onSaveMock}
                    />
                );

                const deleteButton = getByRole('button', { name: /delete/i });
                await user.click(deleteButton);

                expect(simpleWarningModal).toHaveBeenCalledTimes(1);
                expect(simpleWarningModal).toHaveBeenCalledWith(
                    expect.objectContaining({
                        header: 'uniformOffcanvas.deleteAction.header',
                        message: expect.anything(),
                        type: 'danger',
                        primaryOption: 'common.actions.delete',
                        primaryFunction: expect.any(Function),
                    })
                );
                expect(deleteUniformItem).toHaveBeenCalledTimes(0);

                expect(simpleWarningModal.mock.calls[0][0].primaryFunction).toBeDefined();
                await simpleWarningModal.mock.calls[0][0].primaryFunction!();

                expect(deleteUniformItem).toHaveBeenCalledTimes(1);
                expect(deleteUniformItem).toHaveBeenCalledWith(mockUniform.id);
                expect(onCloseMock).toHaveBeenCalledTimes(1);
                expect(onSaveMock).toHaveBeenCalledTimes(1);
                expect(toast.success).toHaveBeenCalledTimes(1);
                expect(toast.success).toHaveBeenCalledWith('uniformOffcanvas.deleteAction.success');
            });
            it('should catch exception', async () => {
                const onCloseMock = jest.fn();
                const onSaveMock = jest.fn();
                deleteUniformItem.mockRejectedValueOnce(new Error('Failed to delete item'));

                const user = userEvent.setup();
                const { getByRole } = render(
                    <UniformOffcanvas
                        uniform={mockUniform}
                        uniformType={typeList[0]}
                        onClose={onCloseMock}
                        onSave={onSaveMock}
                    />
                );

                const deleteButton = getByRole('button', { name: /delete/i });
                await user.click(deleteButton);
                expect(simpleWarningModal).toHaveBeenCalledTimes(1);

                await simpleWarningModal.mock.calls[0][0].primaryFunction!();
                expect(deleteUniformItem).toHaveBeenCalledTimes(1);
                expect(deleteUniformItem).toHaveBeenCalledWith(mockUniform.id);
                expect(onCloseMock).toHaveBeenCalledTimes(0);
                expect(onSaveMock).toHaveBeenCalledTimes(0);
                expect(toast.error).toHaveBeenCalledTimes(1);
                expect(toast.error).toHaveBeenCalledWith('uniformOffcanvas.deleteAction.failed');
            });
        });
    });
    describe('authRoles', () => {
        afterEach(() => delete global.__ROLE__);

        it('should not render edit & delete button if userRole is user', () => {
            global.__ROLE__ = AuthRole.user;
            const { queryByRole } = render(
                <UniformOffcanvas
                    uniform={mockUniform}
                    uniformType={typeList[0]}
                    onClose={jest.fn()}
                    onSave={jest.fn()}
                />
            );

            expect(queryByRole('button', { name: /edit/i })).toBeNull();
            expect(queryByRole('button', { name: /delete/i })).toBeNull();
        });
        it('should render edit & delete button if userRole is higher than user', () => {
            global.__ROLE__ = AuthRole.inspector;
            const { getByRole } = render(
                <UniformOffcanvas
                    uniform={mockUniform}
                    uniformType={typeList[0]}
                    onClose={jest.fn()}
                    onSave={jest.fn()}
                />
            );

            expect(getByRole('button', { name: /edit/i })).toBeInTheDocument();
            expect(getByRole('button', { name: /delete/i })).toBeInTheDocument();
        });

        it('should not render deficiency and history Rows if userRole is user', () => {
            global.__ROLE__ = AuthRole.user;
            const { queryByRole } = render(
                <UniformOffcanvas
                    uniform={mockUniform}
                    uniformType={typeList[0]}
                    onClose={jest.fn()}
                    onSave={jest.fn()}
                />
            );

            expect(queryByRole('list', { name: /deficiency/i })).toBeNull();
            expect(queryByRole('list', { name: /history/i })).toBeNull();
        });
        it('should render deficiency and history Rows if userRole is higher than user', () => {
            global.__ROLE__ = AuthRole.admin;
            const { getByRole } = render(
                <UniformOffcanvas
                    uniform={mockUniform}
                    uniformType={typeList[0]}
                    onClose={jest.fn()}
                    onSave={jest.fn()}
                />
            );

            expect(getByRole('list', { name: /deficiency/i })).toBeInTheDocument();
            expect(getByRole('list', { name: /history/i })).toBeInTheDocument();
        });
    })
});