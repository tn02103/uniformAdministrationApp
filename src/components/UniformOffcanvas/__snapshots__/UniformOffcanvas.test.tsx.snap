// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniformOffcanvas should render correctly 1`] = `
<div
  aria-labelledby="offcanvas-uniform-header"
  aria-modal="true"
  class="offcanvas offcanvas-end offcanvas-toggling show"
  role="dialog"
  style="width: 576px; visibility: visible;"
  tabindex="-1"
>
  <div
    class="offcanvas-header"
  >
    <div
      class="offcanvas-title h5"
    >
      <h3
        id="offcanvas-uniform-header"
      >
        Typ1
        : 
        <b>
          2501
        </b>
      </h3>
    </div>
    <button
      aria-label="Close"
      class="btn-close"
      type="button"
    />
  </div>
  <div
    class="offcanvas-body"
  >
    <h4>
      common.details
    </h4>
    <hr
      class="mb-0"
    />
    <div
      class="justify-content-evenly row"
    >
      <button
        class="col-auto border-0 btn btn-outline-secondary"
        type="button"
      >
        common.actions.edit
         
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-pen fa-sm "
          data-icon="pen"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M362.7 19.3L314.3 67.7 444.3 197.7l48.4-48.4c25-25 25-65.5 0-90.5L453.3 19.3c-25-25-65.5-25-90.5 0zm-71 71L58.6 323.5c-10.4 10.4-18 23.3-22.2 37.4L1 481.2C-1.5 489.7 .8 498.8 7 505s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L421.7 220.3 291.7 90.3z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="col-auto border-0 btn btn-outline-danger"
        type="button"
      >
        common.actions.delete
         
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-trash fa-sm "
          data-icon="trash"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M135.2 17.7L128 32 32 32C14.3 32 0 46.3 0 64S14.3 96 32 96l384 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-96 0-7.2-14.3C307.4 6.8 296.3 0 284.2 0L163.8 0c-12.1 0-23.2 6.8-28.6 17.7zM416 128L32 128 53.2 467c1.6 25.3 22.6 45 47.9 45l245.8 0c25.3 0 46.3-19.7 47.9-45L416 128z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <form
      autocomplete="off"
      class="mb-4"
      novalidate=""
    >
      <div
        class="mb-3 mt-2 justify-content-between row"
      >
        <div
          class="col-sm-auto col-12"
        >
          <div
            class="mb-3"
          >
            <label
              class="fw-bold m-0 form-label"
              for="uniform_toggle-active"
            >
              common.status
            </label>
            <p
              aria-label="common.status"
              aria-readonly="true"
              class="mt-2"
            >
              common.uniform.active.true
            </p>
            <div
              class="text-danger fs-7"
              data-testid="err_active"
              role="alert"
            />
          </div>
        </div>
        <div
          class="col-sm-auto col-7"
        >
          <div
            class="mb-3"
          >
            <label
              class="fw-bold m-0 undefined form-label"
              for="uniform_select-generation"
            >
              common.uniform.generation.label
            </label>
            <p
              aria-label="common.uniform.generation.label"
              aria-readonly="true"
              class="py-2 m-0"
            >
              Generation1-2
            </p>
            <div
              aria-label="error message generation"
              class="text-danger fs-7"
              data-testid="err_generation"
              id="uniform_err_generation"
              role="alert"
            />
          </div>
        </div>
        <div
          class="p-0 col-sm-auto col-5"
          style="min-width: 100px;"
        >
          <div
            class="mb-3"
          >
            <label
              class="fw-bold m-0 undefined form-label"
              for="uniform_select-size"
            >
              common.uniform.size
            </label>
            <p
              aria-label="common.uniform.size"
              aria-readonly="true"
              class="py-2 m-0"
            >
              0
            </p>
            <div
              aria-label="error message size"
              class="text-danger fs-7"
              data-testid="err_size"
              id="uniform_err_size"
              role="alert"
            />
          </div>
        </div>
        <div
          class="col-12"
        >
          <div>
            <label
              class="fw-bold m-0 form-label"
              for="uniform_input-comment"
            >
              common.comment
            </label>
            <textarea
              aria-invalid="false"
              class="form-control-plaintext"
              disabled=""
              id="uniform_input-comment"
              name="comment"
              rows="2"
              width="auto"
            >
              Test comment
            </textarea>
            <div
              aria-label="error message comment"
              class="text-danger fs-7"
              data-testid="err_comment"
              id="uniform_err_comment"
              role="alert"
            />
          </div>
        </div>
      </div>
    </form>
    <h4>
      uniformOffcanvas.deficiency.header
    </h4>
    <hr
      class="mb-0"
    />
    <div
      class="m-0 mb-2 row"
    >
      <div
        class="col"
      >
        <button
          class="mb-2 border-0 btn btn-outline-success"
          type="button"
        >
          common.actions.create
           
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-plus fa-sm "
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 144L48 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l144 0 0 144c0 17.7 14.3 32 32 32s32-14.3 32-32l0-144 144 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-144 0 0-144z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <div
        class="text-end m-2 col-auto"
      >
        <div
          class="form-check form-switch"
        >
          <input
            aria-label="uniformOffcanvas.deficiency.includeResolved"
            class="form-check-input"
            name="showResolved"
            type="checkbox"
          />
          <label
            class="form-check-label"
            title=""
          >
            uniformOffcanvas.deficiency.includeResolved
          </label>
        </div>
      </div>
    </div>
    <div
      aria-label="Deficiency list"
      class="m-0 row"
      role="list"
    >
      <div
        aria-label="uniformOffcanvas.deficiency.cardLabel"
        class="m-1 p-0  card"
        role="listitem"
      >
        <div
          class="position-relative card-body"
        >
          <form
            autocomplete="off"
            class="mb-4"
            novalidate=""
          >
            <div
              class="fs-6 fw-bold card-title h5"
            >
              Test Type
               
            </div>
            <div
              class="position-absolute top-0 end-0"
            >
              <div
                class="dropstart"
              >
                <button
                  aria-expanded="false"
                  aria-label="uniformOffcanvas.deficiency.label.actions"
                  class="border-0 dropdown-toggle btn btn-outline-secondary"
                  id="Cadetdropdown"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-ellipsis-vertical "
                    data-icon="ellipsis-vertical"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 128 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M64 360a56 56 0 1 0 0 112 56 56 0 1 0 0-112zm0-160a56 56 0 1 0 0 112 56 56 0 1 0 0-112zM120 96A56 56 0 1 0 8 96a56 56 0 1 0 112 0z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <p
              class="card-text"
            >
              Test comment
            </p>
            <div
              class="align-items-center justify-content-center row"
              style="height: 2rem;"
            >
              <div
                class="position-relative"
              >
                <hr
                  class=""
                />
                <div
                  class="w-auto bg-white align-center position-absolute top-50 start-50 translate-middle"
                >
                  <button
                    class="border-0 btn btn-light btn-sm"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-chevron-down text-dark me-2 false"
                      data-icon="chevron-down"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                        fill="currentColor"
                      />
                    </svg>
                    expandableArea.showMore
                  </button>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
      <div
        aria-label="uniformOffcanvas.deficiency.cardLabel"
        class="m-1 p-0  card"
        role="listitem"
      >
        <div
          class="position-relative card-body"
        >
          <form
            autocomplete="off"
            class="mb-4"
            novalidate=""
          >
            <div
              class="fs-6 fw-bold card-title h5"
            >
              Another One
               
            </div>
            <div
              class="position-absolute top-0 end-0"
            >
              <div
                class="dropstart"
              >
                <button
                  aria-expanded="false"
                  aria-label="uniformOffcanvas.deficiency.label.actions"
                  class="border-0 dropdown-toggle btn btn-outline-secondary"
                  id="Cadetdropdown"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-ellipsis-vertical "
                    data-icon="ellipsis-vertical"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 128 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M64 360a56 56 0 1 0 0 112 56 56 0 1 0 0-112zm0-160a56 56 0 1 0 0 112 56 56 0 1 0 0-112zM120 96A56 56 0 1 0 8 96a56 56 0 1 0 112 0z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <p
              class="card-text"
            >
              Test comment 2
            </p>
            <div
              class="align-items-center justify-content-center row"
              style="height: 2rem;"
            >
              <div
                class="position-relative"
              >
                <hr
                  class=""
                />
                <div
                  class="w-auto bg-white align-center position-absolute top-50 start-50 translate-middle"
                >
                  <button
                    class="border-0 btn btn-light btn-sm"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-chevron-down text-dark me-2 false"
                      data-icon="chevron-down"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                        fill="currentColor"
                      />
                    </svg>
                    expandableArea.showMore
                  </button>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
    <h4
      class="mt-4"
    >
      uniformOffcanvas.history.header
    </h4>
    <hr
      class="mb-0"
    />
    <div
      class="row"
    >
      <div
        class="m-0 p-4 col"
      >
        <div
          class="bg-light p-2 row"
        >
          <div
            class="fw-bold text-truncate col-4"
          >
            uniformOffcanvas.history.label.dateIssued
            :
          </div>
          <div
            class="fw-bold text-truncate col-4"
          >
            uniformOffcanvas.history.label.dateReturned
            :
          </div>
          <div
            class="fw-bold text-truncate col-4"
          >
            uniformOffcanvas.history.label.cadet
            :
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          >
            01.10.2023
          </div>
          <div
            class="col"
          />
          <div
            class="col"
            title=""
          >
            John
             
            Doe
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          >
            02.10.2023
          </div>
          <div
            class="col"
          >
            03.10.2023
          </div>
          <div
            class="text-decoration-line-through text-danger col"
            title="uniformOffcanvas.history.title.deleted"
          >
            Jane
             
            Smith
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
